Add summary

albertteoh opened the pull request:
## Description

Adds `Summary()` and `String()` functions to output a text summary of each geometry.

## Check List

Have you:

- Added unit tests? Y

- ~Add cmprefimpl tests? (if appropriate?)~

## Related Issue

Fixes #384 and continued on from #393.

peterstace reviewed (commented):
So far this is looking pretty good! I only got through reviewing the geometry collection files, but I'll review the others soon.

albertteoh commented on geom/type_geometry_collection.go:
```diff
@@ -458,3 +459,26 @@ func (c GeometryCollection) DumpCoordinates() Sequence {
+	}
+
+	geometrySuffix := "y"
+	numGeometries := c.NumGeometries()
```
Nice catch! I was thinking, should `NumGeometries` return 7 in this case and  create a new func `NumTopLevelGeometries` to return 1; what do you think?

peterstace replied:
If we go with recursive counting and want to expose that to users (outside of the output of `Summary()`), I think it's better to create a new method (e.g. called `NumTotalGeometries`). There are other `NumFoos` methods on the other types, and they always have corresponding `FooN` methods, essentially allowing users to iterate through the children in a collection in a loop using `for i := 0; i < g.NumFoos(); i++ { child := g.FooN(i); ... }`. To keep geometry collections consistent with the other types, it's nice to keep `NumGeometries` and `GeometryN` as the "length" and "get" functions. The types having this pattern are:

- `MultiPoint` has methods `NumPoints` and `PointN`,
- `MultiLineString` has methods `NumLineStrings` and `LineStringN`,
- `MultiPolygon` has methods `NumPolygons` and `PolygonN`,
- `GeometryCollection` has methods `NumGeometries` and `GeometryN`.

That being said, in the `Summary` output, you _are_ saying "child" geometries, so what you've got right now is correct from that standpoint. It's a fairly weird case to have deeply recursive geometry collections anyway (although it technically can happen). So on a bit more thought, I think what you've already got is ok (although I'm ok with you making it recursive if you want to, I'm happy to leave it up to you).

albertteoh commented on geom/type_geometry_collection_test.go:
```diff
@@ -0,0 +1,78 @@
+		{
+			name: "XY single point",
+			geoms: []geom.Geometry{
+				geom.NewPoint(geom.Coordinates{XY: geom.XY{X: 0, Y: 0}, Type: geom.DimXY}).AsGeometry(),
```
That's much more readable, thanks for pointing this out. I'll see if it makes sense to do the same for the other geometry tests.

peterstace commented on geom/type_polygon.go:
```diff
@@ -2,6 +2,7 @@ package geom
 
 import (
 	"database/sql/driver"
+	"fmt"
```
(commenting here since the file I would comment on for this change isn't in the diff)

The other thing that would be good to add is `Summary() string` and `String() string` on the `Geometry` type. This will unfortunately involve one of those nasty 7-case switch statements (at least until https://github.com/peterstace/simplefeatures/issues/399 is looked at).

This will actually make the testing a little bit easier, since you don't have to call the `Summary` methods directly on each concrete geometry. You can use something like:

```
wkt := "POLYGON((0 0,0 1,1 0,0 0))"
g := geomFromWKT(t, wkt)
got := g.Summary()
// check got vs want
```

Because you're operating on the generic `Geometry` type at that point, you can actually do the Summary tests cases for all 7 types all in the same unit test (since your test cases would just be `string inputWKT` and `string wantSummary`). It may also make sense to put these tests into `geom/attr_test.go`, which contains the other tests that calculate geometry "attributes" (in a sense, the summary string is a calculated "attribute" of a geometry).

albertteoh replied:
i like the suggestion; i noticed I was repeating a very similar pattern on each geometry's test after adopting the WKT approach.

peterstace reviewed (commented):
I finished reviewing the other files, everything is looking pretty good!

albertteoh commented:
I've addressed your comments in [353c2ef3f](https://github.com/peterstace/simplefeatures/pull/400/commits/353c2ef3f96649c159864c060a38c49cfa97659a); PTAL.

peterstace commented:
Perfect! LGTM

albertteoh commented:
Thanks for the review and guidance @peterstace!
